File: src/src/services/calendarService.ts
import axios from 'axios';
import { supabase } from '../lib/supabase';

const GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID;
const GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET;
const GOOGLE_REDIRECT_URI = process.env.GOOGLE_REDIRECT_URI;

const isMockGoogle = !GOOGLE_CLIENT_ID || 
                      GOOGLE_CLIENT_ID === 'seu_google_client_id' || 
                      process.env.MOCK_MODE === 'true';

export interface AppointmentDetails {
  name: string;
  phone: string;
  date: string; // YYYY-MM-DD
  time: string; // HH:MM
  service?: string;
  conversationId?: string;
}

/**
 * Gera a URL do consentimento do Google OAuth para o cliente do AceleraBot.
 */
export function getAuthUrl(clientId: string): string {
  if (isMockGoogle) {
    console.log(`[Mock Google OAuth] Generating login URL for client '${clientId}'`);
    const mockCallback = `${process.env.APP_URL ?? 'http://localhost:3000'}/google/callback?code=mock_authorization_code&state=${clientId}`;
    return mockCallback;
  }

  const scopes = [
    'https://www.googleapis.com/auth/calendar.events',
    'https://www.googleapis.com/auth/calendar.readonly'
  ];

  return `https://accounts.google.com/o/oauth2/v2/auth?` +
    `response_type=code` +
    `&client_id=${GOOGLE_CLIENT_ID}` +
    `&redirect_uri=${encodeURIComponent(GOOGLE_REDIRECT_URI!)}` +
    `&scope=${encodeURIComponent(scopes.join(' '))}` +
    `&access_type=offline` +
    `&prompt=consent` +
    `&state=${clientId}`;
}

/**
 * Troca o código recebido no callback do Google pelos tokens Access e Refresh.
 */
export async function exchangeCodeForTokens(clientId: string, code: string): Promise<void> {
  if (isMockGoogle || code === 'mock_authorization_code') {
    console.log(`[Mock Google OAuth] Exchanging code '${code}' for tokens for client '${clientId}'`);
    const expiryDate = Date.now() + 3600 * 1000; // 1 hora
    
    // Deleta se já existir configuração para evitar conflito de chave única
    await supabase.from('google_calendar_configs').delete().eq('client_id', clientId);

    await supabase.from('google_calendar_configs').insert({
      client_id: clientId,
      access_token: 'mock_access_token_123',
      refresh_token: 'mock_refresh_token_123',
      expiry_date: expiryDate,
      calendar_id: 'primary'
    });
    return;
  }

  try {
    const { data } = await axios.post('https://oauth2.googleapis.com/token', {
      code,
      client_id: GOOGLE_CLIENT_ID,
      client_secret: GOOGLE_CLIENT_SECRET,
      redirect_uri: GOOGLE_REDIRECT_URI,
      grant_type: 'authorization_code',
    });

    const expiryDate = Date.now() + (data.expires_in * 1000);

    // Salva ou atualiza a integração
    await supabase.from('google_calendar_configs').delete().eq('client_id', clientId);

    const { error } = await supabase.from('google_calendar_configs').insert({
      client_id: clientId,
      access_token: data.access_token,
      refresh_token: data.refresh_token,
      expiry_date: expiryDate,
      calendar_id: 'primary',
    });

    if (error) throw error;
  } catch (err: any) {
    console.error('[Google Calendar] Failed to exchange code for tokens:', err.response?.data ?? err.message);
    throw new Error('Falha na autenticação com o Google.');
  }
}

/**
 * Recupera e renova (se necessário) o access token de um cliente.
 */
async function getAccessToken(clientId: string): Promise<{ token: string; calendarId: string } | null> {
  const { data: config } = await supabase
    .from('google_calendar_configs')
    .select('*')
    .eq('client_id', clientId)
    .single();

  if (!config) return null;

  // Se for Mock, retorna as credenciais simuladas
  if (isMockGoogle || config.access_token === 'mock_access_token_123') {
    return { token: config.access_token, calendarId: config.calendar_id };
  }

  const isExpired = Date.now() > (config.expiry_date - 300000); // 5 min de margem

  if (!isExpired) {
    return { token: config.access_token, calendarId: config.calendar_id };
  }

  // Se expirou e temos refresh_token, renova
  if (config.refresh_token) {
    try {
      console.log(`[Google Calendar] Refreshing expired token for client '${clientId}'`);
      const { data } = await axios.post('https://oauth2.googleapis.com/token', {
        client_id: GOOGLE_CLIENT_ID,
        client_secret: GOOGLE_CLIENT_SECRET,
        refresh_token: config.refresh_token,
        grant_type: 'refresh_token',
      });

      const expiryDate = Date.now() + (data.expires_in * 1000);

      await supabase
        .from('google_calendar_configs')
        .update({
          access_token: data.access_token,
          expiry_date: expiryDate,
          updated_at: new Date().toISOString(),
        })
        .eq('client_id', clientId);

      return { token: data.access_token, calendarId: config.calendar_id };
    } catch (err: any) {
      console.error('[Google Calendar] Failed to refresh token:', err.response?.data ?? err.message);
      return null;
    }
  }

  return null;
}

/**
 * Verifica se um cliente tem a integração do Google Calendar ativa.
 */
export async function isGoogleCalendarConnected(clientId: string): Promise<boolean> {
  const { data } = await supabase
    .from('google_calendar_configs')
    .select('id')
    .eq('client_id', clientId)
    .single();
  return !!data;
}

/**
 * Desconecta o Google Calendar removendo a configuração do banco.
 */
export async function disconnectGoogleCalendar(clientId: string): Promise<void> {
  await supabase
    .from('google_calendar_configs')
    .delete()
    .eq('client_id', clientId);
}

/**
 * Retorna os horários livres de um cliente para uma data específica.
 * Combina o expediente comercial do cliente com os horários ocupados obtidos do Google Calendar.
 */
export async function listFreeSlots(clientId: string, dateStr: string): Promise<string[]> {
  const credentials = await getAccessToken(clientId);
  
  // Expediente padrão: Segunda a Sexta, das 08:00 às 18:00 (intervalos de 1 hora)
  // Caso de testes ou sem expediente cadastrado:
  const businessStartHour = 8;
  const businessEndHour = 18;
  const slotDurationMinutes = 60; // Duração dos slots em minutos

  // Se for Mock ou não tiver integração configurada, gera horários livres simulados usando hash da data
  if (!credentials) {
    console.log(`[Google Calendar] Client '${clientId}' is not integrated or in Mock. Simulating free slots for ${dateStr}.`);
    // Simula slots diferentes baseado no dia para parecer dinâmico
    const day = new Date(dateStr + 'T00:00:00').getDay();
    if (day === 0 || day === 6) return []; // Fechado nos finais de semana no mock
    
    // Gera slots com base no dia da semana
    const mockSlots = [];
    for (let hour = businessStartHour; hour < businessEndHour; hour++) {
      if ((hour + day) % 3 !== 0) { // Ocupa alguns horários aleatoriamente
        mockSlots.push(`${hour.toString().padStart(2, '0')}:00`);
      }
    }
    return mockSlots;
  }

  try {
    const timeMin = `${dateStr}T00:00:00Z`;
    const timeMax = `${dateStr}T23:59:59Z`;

    // Chamamos a API do Google Calendar para listar eventos ocupados naquele dia
    const response = await axios.get(
      `https://www.googleapis.com/calendar/v3/calendars/${credentials.calendarId}/events`,
      {
        headers: { Authorization: `Bearer ${credentials.token}` },
        params: {
          timeMin,
          timeMax,
          singleEvents: true,
          orderBy: 'startTime'
        }
      }
    );

    const events = response.data.items ?? [];
    const busyRanges = events.map((event: any) => {
      const start = new Date(event.start.dateTime ?? event.start.date);
      const end = new Date(event.end.dateTime ?? event.end.date);
      return { start, end };
    });

    // Gera todos os slots possíveis dentro do expediente
    const freeSlots: string[] = [];
    const dateObj = new Date(dateStr + 'T00:00:00');

    // Se final de semana, não há vagas
    if (dateObj.getDay() === 0 || dateObj.getDay() === 6) {
      return [];
    }

    for (let hour = businessStartHour; hour < businessEndHour; hour++) {
      const slotTimeStr = `${hour.toString().padStart(2, '0')}:00`;
      const slotStart = new Date(`${dateStr}T${slotTimeStr}:00`);
      const slotEnd = new Date(slotStart.getTime() + slotDurationMinutes * 60 * 1000);

      // Verifica se o slot de horário colide com algum compromisso
      const isBusy = busyRanges.some((busy: any) => {
        return (slotStart < busy.end && slotEnd > busy.start);
      });

      if (!isBusy) {
        freeSlots.push(slotTimeStr);
      }
    }

    return freeSlots;
  } catch (err: any) {
    console.error('[Google Calendar] Failed to fetch freeBusy / events:', err.response?.data ?? err.message);
    // Em caso de erro, retorna vazio
    return [];
  }
}

/**
 * Cria um agendamento no banco de dados e adiciona o evento no Google Calendar.
 */
export async function createCalendarEvent(clientId: string, details: AppointmentDetails): Promise<any> {
  const credentials = await getAccessToken(clientId);
  
  // Duração padrão de 1 hora
  const startISO = `${details.date}T${details.time}:00`;
  const startDate = new Date(startISO);
  const endDate = new Date(startDate.getTime() + 60 * 60 * 1000);
  const endISO = endDate.toISOString();

  let googleEventId: string | null = null;

  if (credentials) {
    try {
      // Chama Google Calendar API para agendar
      const response = await axios.post(
        `https://www.googleapis.com/calendar/v3/calendars/${credentials.calendarId}/events`,
        {
          summary: `Agendamento: ${details.name}`,
          description: `Serviço: ${details.service ?? 'Geral'}\nTelefone: +${details.phone}\nAgendado pelo AceleraAssistente 🤖`,
          start: {
            dateTime: startDate.toISOString(),
            timeZone: 'America/Sao_Paulo'
          },
          end: {
            dateTime: endDate.toISOString(),
            timeZone: 'America/Sao_Paulo'
          }
        },
        {
          headers: { Authorization: `Bearer ${credentials.token}` }
        }
      );
      googleEventId = response.data.id;
    } catch (err: any) {
      console.error('[Google Calendar] Failed to create event in Google:', err.response?.data ?? err.message);
      // Salva no banco de dados local de qualquer forma
    }
  } else {
    // Modo Mock
    googleEventId = 'mock_evt_' + Math.random().toString(36).substring(7);
    console.log(`[Mock Google Calendar] Created event '${googleEventId}' for ${details.name} at ${startISO}`);
  }

  // Registra no banco de dados local
  const { data: appointment, error } = await supabase
    .from('appointments')
    .insert({
      client_id: clientId,
      conversation_id: details.conversationId ?? null,
      lead_phone: details.phone,
      lead_name: details.name,
      service_name: details.service ?? 'Geral',
      start_time: startDate.toISOString(),
      end_time: endDate.toISOString(),
      google_event_id: googleEventId,
      status: 'scheduled'
    })
    .select()
    .single();

  if (error) {
    console.log('[Google Calendar] Failed to save appointment in Supabase:', error);
    throw error;
  }

  return appointment;
}

/**
 * Exclui um evento do Google Calendar e do banco de dados (cancela).
 */
export async function deleteCalendarEvent(clientId: string, googleEventId: string): Promise<void> {
  const credentials = await getAccessToken(clientId);

  if (credentials && googleEventId && !googleEventId.startsWith('mock_')) {
    try {
      await axios.delete(
        `https://www.googleapis.com/calendar/v3/calendars/${credentials.calendarId}/events/${googleEventId}`,
        {
          headers: { Authorization: `Bearer ${credentials.token}` }
        }
      );
      console.log(`[Google Calendar] Deleted event '${googleEventId}' from Google.`);
    } catch (err: any) {
      console.error('[Google Calendar] Failed to delete event in Google:', err.response?.data ?? err.message);
      // Continua para cancelar localmente mesmo se falhar no Google
    }
  } else {
    console.log(`[Mock Google Calendar] Deleted event '${googleEventId}' from Mock Calendar.`);
  }
}

File: AceleraBot.Api/Services/CalendarService.cs
using System.Text.Json;
using AceleraBot.Api.Data;
using AceleraBot.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace AceleraBot.Api.Services;

public interface ICalendarService
{
    string GetAuthUrl(Guid clientId);
    Task ExchangeCodeForTokensAsync(Guid clientId, string code);
    Task<List<string>> ListFreeSlotsAsync(Guid clientId, string dateStr);
    Task<Appointment> CreateCalendarEventAsync(Guid clientId, AppointmentInfo info, string phone, Guid? conversationId);
    Task DeleteCalendarEventAsync(Guid clientId, string googleEventId);
}

// Espelha src/services/calendarService.ts.
public class CalendarService : ICalendarService
{
    private readonly AppDbContext _db;
    private readonly IHttpClientFactory _factory;
    private readonly IConfiguration _cfg;
    private readonly ILogger<CalendarService> _log;
    private readonly bool _isMock;
    private static readonly TimeZoneInfo Tz = ResolveTz();

    public CalendarService(AppDbContext db, IHttpClientFactory factory, IConfiguration cfg, ILogger<CalendarService> log)
    {
        _db = db; _factory = factory; _cfg = cfg; _log = log;
        var id = cfg["GOOGLE_CLIENT_ID"];
        _isMock = string.IsNullOrEmpty(id) || id == "seu_google_client_id" || cfg["MOCK_MODE"] == "true";
    }

    private static TimeZoneInfo ResolveTz()
    {
        try { return TimeZoneInfo.FindSystemTimeZoneById("America/Sao_Paulo"); }
        catch { return TimeZoneInfo.Utc; }
    }

    public string GetAuthUrl(Guid clientId)
    {
        if (_isMock)
            return $"{_cfg["APP_URL"] ?? "http://localhost:3000"}/google/callback?code=mock_authorization_code&state={clientId}";

        var scopes = "https://www.googleapis.com/auth/calendar.events https://www.googleapis.com/auth/calendar.readonly";
        var redirect = _cfg["GOOGLE_REDIRECT_URI"] ?? "";
        return "https://accounts.google.com/o/oauth2/v2/auth?response_type=code"
             + $"&client_id={_cfg["GOOGLE_CLIENT_ID"]}"
             + $"&redirect_uri={Uri.EscapeDataString(redirect)}"
             + $"&scope={Uri.EscapeDataString(scopes)}"
             + "&access_type=offline&prompt=consent"
             + $"&state={clientId}";
    }

    public async Task ExchangeCodeForTokensAsync(Guid clientId, string code)
    {
        string accessToken, refreshToken; long expiryDate;
        if (_isMock)
        {
            accessToken = "mock_access_token"; refreshToken = "mock_refresh_token";
            expiryDate = DateTimeOffset.UtcNow.AddHours(1).ToUnixTimeMilliseconds();
        }
        else
        {
            var form = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["code"] = code,
                ["client_id"] = _cfg["GOOGLE_CLIENT_ID"] ?? "",
                ["client_secret"] = _cfg["GOOGLE_CLIENT_SECRET"] ?? "",
                ["redirect_uri"] = _cfg["GOOGLE_REDIRECT_URI"] ?? "",
                ["grant_type"] = "authorization_code",
            });
            var res = await _factory.CreateClient().PostAsync("https://oauth2.googleapis.com/token", form);
            res.EnsureSuccessStatusCode();
            var json = await res.Content.ReadFromJsonAsync<JsonElement>();
            accessToken = json.GetProperty("access_token").GetString()!;
            refreshToken = json.TryGetProperty("refresh_token", out var rt) ? rt.GetString() ?? "" : "";
            var expiresIn = json.TryGetProperty("expires_in", out var ei) ? ei.GetInt32() : 3600;
            expiryDate = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (long)expiresIn * 1000;
        }

        var existing = await _db.GoogleCalendarConfigs.Where(c => c.ClientId == clientId).ToListAsync();
        _db.GoogleCalendarConfigs.RemoveRange(existing);
        _db.GoogleCalendarConfigs.Add(new GoogleCalendarConfig
        {
            ClientId = clientId,
            AccessToken = accessToken,
            RefreshToken = refreshToken,
            ExpiryDate = expiryDate,
            CalendarId = "primary",
        });
        await _db.SaveChangesAsync();
    }

    private async Task<(string token, string calendarId)?> GetAccessTokenAsync(Guid clientId)
    {
        var cfg = await _db.GoogleCalendarConfigs.FirstOrDefaultAsync(c => c.ClientId == clientId);
        if (cfg is null) return null;
        if (_isMock) return (cfg.AccessToken, cfg.CalendarId);

        var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (cfg.ExpiryDate is not null && nowMs > cfg.ExpiryDate.Value - 5 * 60 * 1000 && !string.IsNullOrEmpty(cfg.RefreshToken))
        {
            var form = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["client_id"] = _cfg["GOOGLE_CLIENT_ID"] ?? "",
                ["client_secret"] = _cfg["GOOGLE_CLIENT_SECRET"] ?? "",
                ["refresh_token"] = cfg.RefreshToken!,
                ["grant_type"] = "refresh_token",
            });
            var res = await _factory.CreateClient().PostAsync("https://oauth2.googleapis.com/token", form);
            if (res.IsSuccessStatusCode)
            {
                var json = await res.Content.ReadFromJsonAsync<JsonElement>();
                cfg.AccessToken = json.GetProperty("access_token").GetString()!;
                var expiresIn = json.TryGetProperty("expires_in", out var ei) ? ei.GetInt32() : 3600;
                cfg.ExpiryDate = nowMs + (long)expiresIn * 1000;
                await _db.SaveChangesAsync();
            }
        }
        return (cfg.AccessToken, cfg.CalendarId);
    }

    public async Task<List<string>> ListFreeSlotsAsync(Guid clientId, string dateStr)
    {
        var businessHours = Enumerable.Range(8, 10).Select(h => $"{h:D2}:00").ToList(); // 08:00..17:00
        var creds = await GetAccessTokenAsync(clientId);

        if (creds is null)
        {
            // Simulação (sem credenciais): fim de semana vazio, senão alguns horários
            if (DateTime.TryParse(dateStr, out var d) && (d.DayOfWeek == DayOfWeek.Saturday || d.DayOfWeek == DayOfWeek.Sunday))
                return new List<string>();
            return new List<string> { "09:00", "11:00", "14:00", "16:00" };
        }

        try
        {
            var client = _factory.CreateClient();
            client.DefaultRequestHeaders.Authorization = new("Bearer", creds.Value.token);
            var url = $"https://www.googleapis.com/calendar/v3/calendars/{Uri.EscapeDataString(creds.Value.calendarId)}/events"
                    + $"?timeMin={dateStr}T00:00:00Z&timeMax={dateStr}T23:59:59Z&singleEvents=true&orderBy=startTime";
            var res = await client.GetAsync(url);
            if (!res.IsSuccessStatusCode) return businessHours;
            var json = await res.Content.ReadFromJsonAsync<JsonElement>();
            var busyHours = new HashSet<int>();
            if (json.TryGetProperty("items", out var items))
            {
                foreach (var ev in items.EnumerateArray())
                {
                    if (ev.TryGetProperty("start", out var s) && s.TryGetProperty("dateTime", out var sdt)
                        && DateTimeOffset.TryParse(sdt.GetString(), out var start))
                        busyHours.Add(start.Hour);
                }
            }
            return businessHours.Where(h => !busyHours.Contains(int.Parse(h[..2]))).ToList();
        }
        catch (Exception e) { _log.LogError(e, "listFreeSlots failed"); return businessHours; }
    }

    public async Task<Appointment> CreateCalendarEventAsync(Guid clientId, AppointmentInfo info, string phone, Guid? conversationId)
    {
        var startLocal = DateTime.Parse($"{info.Date}T{info.Time}:00");
        var startUtc = TimeZoneInfo.ConvertTimeToUtc(DateTime.SpecifyKind(startLocal, DateTimeKind.Unspecified), Tz);
        var endUtc = startUtc.AddHours(1);
        string? googleEventId = null;

        var creds = await GetAccessTokenAsync(clientId);
        if (creds is not null && !_isMock)
        {
            try
            {
                var client = _factory.CreateClient();
                client.DefaultRequestHeaders.Authorization = new("Bearer", creds.Value.token);
                var body = new
                {
                    summary = $"Agendamento: {info.Name}",
                    description = $"Serviço: {info.Service}\nTelefone: +{phone}\nAgendado pelo AceleraAssistente 🤖",
                    start = new { dateTime = startLocal.ToString("yyyy-MM-ddTHH:mm:ss"), timeZone = "America/Sao_Paulo" },
                    end = new { dateTime = startLocal.AddHours(1).ToString("yyyy-MM-ddTHH:mm:ss"), timeZone = "America/Sao_Paulo" },
                };
                var url = $"https://www.googleapis.com/calendar/v3/calendars/{Uri.EscapeDataString(creds.Value.calendarId)}/events";
                var res = await client.PostAsJsonAsync(url, body);
                if (res.IsSuccessStatusCode)
                {
                    var json = await res.Content.ReadFromJsonAsync<JsonElement>();
                    googleEventId = json.TryGetProperty("id", out var idv) ? idv.GetString() : null;
                }
            }
            catch (Exception e) { _log.LogError(e, "createCalendarEvent (google) failed"); }
        }

        var appt = new Appointment
        {
            ClientId = clientId,
            ConversationId = conversationId,
            LeadPhone = phone,
            LeadName = info.Name,
            ServiceName = info.Service,
            StartTime = startUtc,
            EndTime = endUtc,
            GoogleEventId = googleEventId,
            Status = "scheduled",
        };
        _db.Appointments.Add(appt);
        await _db.SaveChangesAsync();
        return appt;
    }

    public async Task DeleteCalendarEventAsync(Guid clientId, string googleEventId)
    {
        if (_isMock || string.IsNullOrEmpty(googleEventId) || googleEventId.StartsWith("mock")) return;
        var creds = await GetAccessTokenAsync(clientId);
        if (creds is null) return;
        try
        {
            var client = _factory.CreateClient();
            client.DefaultRequestHeaders.Authorization = new("Bearer", creds.Value.token);
            await client.DeleteAsync($"https://www.googleapis.com/calendar/v3/calendars/{Uri.EscapeDataString(creds.Value.calendarId)}/events/{googleEventId}");
        }
        catch (Exception e) { _log.LogError(e, "deleteCalendarEvent failed"); }
    }
}
